src/frontend/src/expr/utils.rs

```diff
@@ -229,6 +241,47 @@ impl ExprRewriter for NotPushDown {
     }
 }
 
+pub fn extract_common_factor(expr: ExprImpl) -> Vec<ExprImpl> {
+    // For example, we assume that expr is (A & B & C & D) | (B & C & D) | (C & D & E)
+    let disjunctions: Vec<ExprImpl> = to_disjunctions(expr);
+
+    // if `expr` can not be divided into several disjunctions...
+    if disjunctions.len() == 1 {
+        return disjunctions;
+    }
+
+    // extract (A & B & C & D) | (B & C & D) | (C & D & E)
+    // into [[A, B, C, D], [B, C, D], [C, D, E]]
+    let mut disjunctions: Vec<Vec<_>> = disjunctions
+        .into_iter()
+        .map(|x| to_conjunctions(x).into_iter().collect())
+        .collect();
+    let last = disjunctions.pop().unwrap();
+    let (greatest_common_divider, others): (Vec<_>, _) = last
+        .into_iter()
+        .partition(|x| disjunctions.iter().all(|y| y.contains(x)));
+    // now greatest_common_factor == [C, D]
+    disjunctions.push(others);
```

Comment:
What about `last_mut` + `drain_filter`?

Reply:
`drain_filter` is nightly and requires a special feature gate, and I add it to `lib.rs`. 

`last_mut` might not be used, because when `drain_filter`ing the last element of `disjunctions`, `disjunctions` itself will be borrowed as well. Thus we have an immutable borrow and a mutable borrow(`last_mut`) to `disjunctions`. The compiler 🦀 is not happy with that :rofl: .